scalyr_agent/util.py

```diff
@@ -2434,3 +2436,226 @@ def match_glob(pathname):
         # use the third party glob library to handle a recursive glob.
         result = glob2.glob(pathname)
     return result
+
+
+class ProcessWatchDog(threading.Thread):
+    """
+    A watchdog that creates a new thread and pools a particular process.
+    If the process does not exist anymore, the watchdog invokes a callback.
+    """
+
+    def __init__(self, pid, poll_interval):
+        """
+        :param pid: target process PID
+        :param poll_interval: The polling interval in seconds.
+        """
+        super(ProcessWatchDog, self).__init__(name="watchdog thread.")
+
+        self.daemon = True
+
+        self.pid = pid
+        self._poll_interval = poll_interval
+
+        # the callback which is invoked when the process iskilled.
+        self._on_stop_callback = None
+
+    def run(self):
+        """
+        Start probing the process.
+        :return:
+        """
+        while True:
+            try:
+                # probing the parent process.
+                if self.pid is not None:
+                    os.kill(self.pid, 0)
+                time.sleep(self._poll_interval)
+            except OSError:
+                self._on_stop_callback()
```

Comment:
For double protection, I think we should also wait X seconds here and then send a KILL signal to our own process.  I know the main agent process is also set to kill this process, but it is safest to do that here as well.

So, be sure to catch all exceptions when you invoke `self._on_stop_callback` and then sleep for 10 seconds, and then KILL this process.

Reply:
Let's have the kill time also be an instance variable set like the poll interval.

Reply:
In the case of our manager class, its member function `ParentProcessAwareSyncManager._terminate` is set as `_on_stop_callback` for a watchdog, so I guess it would be better to do a kill there and keep the watchdog class "less specialized".
https://github.com/scalyr/scalyr-agent-2/blob/3586c46d34bddf17aec1191ea173115e87213417/scalyr_agent/util.py#L2587-L2607